submissions/morenko/task-frogger/js/app.js

```diff
@@ -0,0 +1,177 @@
+        this.minX = 0;
+    }
+
+    reset() {
+        this.x = this.startX;
+        this.y = this.startY;
+        
+    }
+}
+
+class Enemy {
+
+    speed = (Math.floor(Math.random(4) * (500 - 100 + 1)) + 100);
+    sprite = 'images/Rock.png';
+
+    constructor ( PositionEnemy ) {
+        this.position = PositionEnemy;
+    }
+
+    update (dt) {
+        if (this.position.x <= this.position.maxX) {
+            this.position.run(this.speed, dt);
+        } else {
+            this.position.reset();
+        }
+    }
+
+    render = function() {
+        ctx.drawImage(Resources.get(this.sprite), this.position.x, this.position.y);
+    }
+}
+
+class PositionEnemy extends Position {
+
+    run(speed, dt) {
+        this.x += (speed * dt); 
+    }
+}
+
+class Player {
+    sprite = 'images/char-princess-girl.png';
+
+    constructor ( PositionPlayer ) {
+        this.position = PositionPlayer;
+    }
+
+    alertWin() {
+        if (this.position.y == this.position.minY) {
+            setTimeout(() => {
+                alert("Not bed! Try again!");
+                this.position.reset();
+            }, 0);
+        }
+    }
+
+    update () {
+        this.alertWin();
+    }
+
+    handleInput (keyCode) {
+
+        switch(keyCode) {
+            case 'up': 
+                this.position.moveUp();
+                break;
+            case 'down': 
+                this.position.moveDown();
+                break;
+            case 'right': 
+                this.position.moveRigtht();
+                break;
+            case 'left': 
+                this.position.moveLeft();
+                break;
+            default: 
+                break;
+        }
+    }
+
+    render () {
+        ctx.drawImage(Resources.get(this.sprite), this.position.x, this.position.y);
+    }
+}
+
+class PositionPlayer extends Position {
+
+    moveRigtht() {
+        let newPositionX = this.x + tileWidth;
+        if (newPositionX <= this.maxX) {
+            this.x = newPositionX; 
+        }
+    }
+
+    moveLeft() {
+        let newPositionX = this.x - tileWidth;
+        if (newPositionX >= this.minX) {
+            this.x = newPositionX; 
+        }
+    }
+
+    moveUp() {
+        let newPositionY = this.y - tileHight;
+        if (newPositionY >= this.minY) {
+            this.y = newPositionY; 
+        }
+    }
+
+    moveDown() {
+        let newPositionY = this.y + tileHight;
+        if (newPositionY <= this.maxY) {
+            this.y = newPositionY;
+        }
+    }
+}
+
+const player = new Player (new PositionPlayer( 5, 3 ));
+
+const allEnemies = [
+    new Enemy (new PositionEnemy ( 1, 0 )), 
+    new Enemy (new PositionEnemy ( 2, 0 )),
+    new Enemy (new PositionEnemy ( 3, 0 )),
+];
+
+function checkCollisions() {
+    const confrontEnemies = allEnemies.filter(filterConfrontEnemies, player);
+    
+    if (confrontEnemies.length) {
+        player.position.reset();
+    }
+}
+
+function filterConfrontEnemies(enemy) {
```

Comment:
This function could be confusing. This is a clever usage of `this` argument, but it's not very practical. We can make this function a useful tool that checks for collision between 2 objects. It's better to just pass a second object as a second argument.